File: protocol-generator/java/src/main/java/org/entitycore/protocol/codec/EcfValue.java
package org.entitycore.protocol.codec;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;
import java.util.List;
import java.util.Objects;

/**
 * The decoded-form value model for Entity Canonical Form (ECF).
 *
 * <p>A {@code sealed interface} + nested records gives the codec/dispatch ladders an
 * EXHAUSTIVE pattern-matching switch (JDK 21 JEP 441) over a closed type set — the
 * static-OO analogue of the OCaml variant / Zig tagged union / CL sentinel approach.
 *
 * <p>Why an explicit model rather than reusing {@code java.lang} types directly:
 * ECF requires {@code absent != null != false != 0} on the wire (V7 §1.3), and a
 * CBOR byte string (major 2) must stay distinct from a text string (major 3). So:
 * <ul>
 *   <li>booleans/null are explicit nodes ({@link Bool}, {@link Null}) — never erased
 *       to a Java {@code null}/{@code boolean};</li>
 *   <li>byte strings are {@link Bytes} (major 2), text is {@link Text} (major 3) —
 *       never conflated;</li>
 *   <li>integral-valued floats keep an explicit {@link Float64} node so {@code 1.0}
 *       encodes as a float ({@code f90000} ladder), never as integer 1 (the
 *       value-erasure trap TS hit with cborg);</li>
 *   <li>the float specials NaN/±Inf/-0.0 are carried as {@link FloatSpecial}
 *       constants so a NaN/Inf never has to round-trip through a Java {@code double}.</li>
 * </ul>
 *
 * <p>Integers use {@code java.math.BigInteger} ({@link Int}) so the full uint64 /
 * -2^64 head-form range is representable with NO native-unsigned trap (Java
 * {@code long} has no native unsigned; this side-steps it entirely).
 */
public sealed interface EcfValue
        permits EcfValue.Int, EcfValue.Float64, EcfValue.FloatSpecial,
                EcfValue.Bytes, EcfValue.Text, EcfValue.Array, EcfValue.Map,
                EcfValue.Bool, EcfValue.Null {

    /** CBOR major-type 0/1 integer, as a {@link java.math.BigInteger} (full range). */
    record Int(java.math.BigInteger value) implements EcfValue {
        public Int {
            Objects.requireNonNull(value, "value");
        }
        public static Int of(long v) { return new Int(java.math.BigInteger.valueOf(v)); }
    }

    /** A finite floating-point value (encoded via the shortest f16/f32/f64 ladder). */
    record Float64(double value) implements EcfValue { }

    /** The ECF Rule-4a special floats. Carried as sentinels so the canonical wire
     *  bytes are emitted directly and a NaN/Inf never materializes as a {@code double}. */
    enum FloatSpecial implements EcfValue {
        NAN, POSITIVE_INFINITY, NEGATIVE_INFINITY, NEGATIVE_ZERO
    }

    /** CBOR byte string (major 2). The {@code byte[]} is defensively copied in and
     *  out — Java arrays are mutable, so the codec NEVER aliases an internal buffer
     *  (profile {@code no_byte_array_aliasing}). */
    record Bytes(byte[] octets) implements EcfValue {
        public Bytes(byte[] octets) {
            this.octets = octets.clone();
        }
        /** Defensive copy on read. */
        @Override public byte[] octets() { return octets.clone(); }
        /** Internal, no-copy accessor for the codec hot path (caller must not mutate). */
        byte[] raw() { return octets; }
        @Override public boolean equals(Object o) {
            return o instanceof Bytes b && Arrays.equals(octets, b.octets);
        }
        @Override public int hashCode() { return Arrays.hashCode(octets); }
        @Override public String toString() { return "Bytes(" + octets.length + "B)"; }
    }

    /** CBOR text string (major 3), held as a Java {@link String} (UTF-8 on the wire). */
    record Text(String value) implements EcfValue {
        public Text {
            Objects.requireNonNull(value, "value");
        }
    }

    /** CBOR array (major 4), definite length. Order-preserving, immutable view. */
    record Array(List<EcfValue> items) implements EcfValue {
        public Array(List<EcfValue> items) {
            this.items = Collections.unmodifiableList(new ArrayList<>(items));
        }
        public static Array of(EcfValue... items) { return new Array(Arrays.asList(items)); }
    }

    /** CBOR map (major 5). Order is whatever was constructed/decoded; the ENCODER
     *  re-sorts by encoded-key length-then-lex (ECF Rule 2). Keys are {@link Text}
     *  or {@link Bytes}. Held as an ordered list of {@link Entry} so decode order is
     *  preserved and duplicate detection is the decoder's job (not a Map collision). */
    record Map(List<Entry> entries) implements EcfValue {
        public Map(List<Entry> entries) {
            this.entries = Collections.unmodifiableList(new ArrayList<>(entries));
        }
        /** A single map key/value pair (key is Text or Bytes). */
        public record Entry(EcfValue key, EcfValue value) { }

        /** Build a map from alternating key, value, key, value... (Text keys via String). */
        public static Map of(Object... kvs) {
            if ((kvs.length & 1) != 0) {
                throw new IllegalArgumentException("odd kv count");
            }
            List<Entry> es = new ArrayList<>(kvs.length / 2);
            for (int i = 0; i < kvs.length; i += 2) {
                EcfValue k = (kvs[i] instanceof String s) ? new Text(s) : (EcfValue) kvs[i];
                es.add(new Entry(k, (EcfValue) kvs[i + 1]));
            }
            return new Map(es);
        }

        /** Fetch a value by Text key; null if absent. */
        public EcfValue get(String key) {
            for (Entry e : entries) {
                if (e.key() instanceof Text t && t.value().equals(key)) {
                    return e.value();
                }
            }
            return null;
        }
    }

    /** CBOR true/false (major 7). */
    enum Bool implements EcfValue { TRUE, FALSE }

    /** CBOR null (major 7, value 22). A singleton — distinct from a Java null. */
    enum Null implements EcfValue { NULL }
}

File: protocol-generator/kotlin/src/main/kotlin/org/entitycore/protocol/codec/EcfValue.kt
package org.entitycore.protocol.codec

import java.math.BigInteger

/**
 * The decoded-form value model for Entity Canonical Form (ECF).
 *
 * A `sealed interface` hierarchy gives the codec/dispatch ladders an EXHAUSTIVE `when`
 * (the compiler checks coverage over the closed set — no `else` needed). This is the
 * Kotlin static-exhaustiveness seam. A sealed *interface* (not class) is used so the
 * value-less variants ([Bool], [FloatSpecial], [Null]) can be Kotlin `enum`/`object`
 * singletons and still be first-class `EcfValue`s in the `when` (Kotlin enums/objects
 * implement interfaces; this is the idiomatic closed-hierarchy shape here).
 *
 * Why an explicit model rather than reusing `kotlin` stdlib types directly: ECF
 * requires `absent != null != false != 0` on the wire (V7 §1.3), and a CBOR byte string
 * (major 2) must stay distinct from a text string (major 3). So:
 *  - booleans/null are explicit nodes ([Bool], [Null]) — never erased to a Kotlin
 *    `null`/`Boolean`;
 *  - byte strings are [Bytes] (major 2), text is [Text] (major 3) — never conflated;
 *  - integral-valued floats keep an explicit [Float64] node so `1.0` encodes as a
 *    float (the `f90000` ladder), never as integer 1;
 *  - the float specials NaN/±Inf/-0.0 are carried as [FloatSpecial] constants so a
 *    NaN/Inf never has to round-trip through a Kotlin `Double`.
 *
 * Integers use [BigInteger] ([IntVal]) so the full uint64 / -2^64 head-form range is
 * representable. Kotlin's `ULong` (profile [idiom] unsigned_types) carries the unsigned
 * uint64 head-form natively without the signedness-comparison footgun — the win shows up
 * in the head-FORM emitter (see [CanonicalCbor], which switches argument length on a
 * `ULong`), while the value NODE keeps `BigInteger` because the nint side reaches -2^64
 * (one past `ULong`).
 */
sealed interface EcfValue {

    /** CBOR major-type 0/1 integer, as a [BigInteger] (full uint64 / -2^64 range). */
    data class IntVal(val value: BigInteger) : EcfValue {
        companion object {
            fun of(v: Long): IntVal = IntVal(BigInteger.valueOf(v))
        }
    }

    /** A finite floating-point value (encoded via the shortest f16/f32/f64 ladder). */
    data class Float64(val value: Double) : EcfValue

    /** The ECF Rule-4a special floats. Carried as sentinels so the canonical wire bytes
     *  are emitted directly and a NaN/Inf never materializes as a `Double`. */
    enum class FloatSpecial : EcfValue {
        NAN, POSITIVE_INFINITY, NEGATIVE_INFINITY, NEGATIVE_ZERO
    }

    /** CBOR byte string (major 2). The [ByteArray] is defensively copied in and out —
     *  JVM `ByteArray` is mutable + aliasable (profile no_byte_array_aliasing); a `val`
     *  does NOT make the contents immutable, so the codec NEVER aliases an internal
     *  buffer by reference. */
    class Bytes(octets: ByteArray) : EcfValue {
        private val data: ByteArray = octets.copyOf()
        /** Defensive copy on read. */
        fun octets(): ByteArray = data.copyOf()
        /** Internal no-copy accessor for the codec hot path (caller must not mutate). */
        internal fun raw(): ByteArray = data
        override fun equals(other: Any?): Boolean =
            other is Bytes && data.contentEquals(other.data)
        override fun hashCode(): Int = data.contentHashCode()
        override fun toString(): String = "Bytes(${data.size}B)"
    }

    /** CBOR text string (major 3), held as a Kotlin [String] (UTF-8 on the wire). */
    data class Text(val value: String) : EcfValue

    /** CBOR array (major 4), definite length. Read-only view on a defensive copy. */
    class Arr(items: List<EcfValue>) : EcfValue {
        val items: List<EcfValue> = items.toList()
        override fun equals(other: Any?): Boolean = other is Arr && items == other.items
        override fun hashCode(): Int = items.hashCode()
        companion object {
            fun of(vararg items: EcfValue): Arr = Arr(items.toList())
        }
    }

    /** A single map key/value pair (key is [Text] or [Bytes]). */
    data class Entry(val key: EcfValue, val value: EcfValue)

    /** CBOR map (major 5). Decode/construct order is preserved as a list of [Entry];
     *  the ENCODER re-sorts by encoded-key length-then-lex (ECF Rule 2). Keys are
     *  [Text] or [Bytes]. */
    class MapVal(entries: List<Entry>) : EcfValue {
        val entries: List<Entry> = entries.toList()

        /** Fetch a value by Text key; null if absent. */
        operator fun get(key: String): EcfValue? =
            entries.firstOrNull { (it.key as? Text)?.value == key }?.value

        override fun equals(other: Any?): Boolean = other is MapVal && entries == other.entries
        override fun hashCode(): Int = entries.hashCode()

        companion object {
            /** Build a map from alternating key, value... (a String key is wrapped as Text). */
            fun of(vararg kvs: Any?): MapVal {
                require(kvs.size % 2 == 0) { "odd kv count" }
                val es = ArrayList<Entry>(kvs.size / 2)
                var i = 0
                while (i < kvs.size) {
                    val k = kvs[i]
                    val key = if (k is String) Text(k) else k as EcfValue
                    es.add(Entry(key, kvs[i + 1] as EcfValue))
                    i += 2
                }
                return MapVal(es)
            }
        }
    }

    /** CBOR true/false (major 7). */
    enum class Bool : EcfValue { TRUE, FALSE }

    /** CBOR null (major 7, value 22). A singleton — distinct from a Kotlin null. */
    object Null : EcfValue
}
